Connection is now two-way

khoin opened the pull request:
Client-side now prints message sent from server.

If you look in `cmd.go`, you'll see that anything `cmd.Println` gets fed will be printed on the server-side. I believe we're not limited just to text but also any other kind of data through output ioWriter, exposed by Cobra: https://github.com/spf13/cobra/blob/master/command.go#L233-L236

You can still print to the console, through `fmt`. Look into `cmd.go`, for example: When the user requests for `dummy version`. The local console will output "I'm version!" and the user receives "Version 1".

The instantiation of `gobra.Server` now requires a `*cobra.Command` as the first argument as well.

The next step will be resolving flags as well as arguments, etc. Let me know what you think. Currently, I think we're safe, as this only executes whatever that's in `cmd.go`. If the user tampers with the client-side (i.e., trying to run a command that's not in cmd.go), we will just run the Root command. This is because whenever the server receives the list of command/sub-command, it uses `cobra.Find`, and `cobra.Find` returns the Root command is the command isn't found in the tree.

Since I'm using `cobra.Command.Run()` instead of `cobra.Command.Execute()`, PreRun, PersistentRun aren't appear to be running, yet. I'll look into that as well. 

If this PR isn't sufficient, let me know and I'll lump it until more things are settled.

Commit:
Connection is now two-way

Fixes made

ctessum commented on example/cmd/cmd.go:
```diff
@@ -43,8 +44,9 @@ var Root = &cobra.Command{
-	PersistentPreRun: func(cmd *cobra.Command, args []string) {
-		fmt.Println(`I'm ran.`)
+	Run: func(cmd *cobra.Command, args []string) {
+		fmt.Println(`I'm ran!`)
```
"I'm dummy"?

ctessum reviewed (approved):
Looks good. Just one change. This is a fine amount for one pull request.

For next steps, in addition to the pre-run and post-run functions, we will also want the user to be able to set the flags. An example program could be one where the user specifies two numbers (as flags) in the html UI and then executes the command, which would add the two numbers together and print the result back to the client.

ctessum commented:
Also, is there a way to use the commands to run the server using a script or command line rather than the web ui? If so, would you be able to create an example using curl or python or go or something like that?

khoin commented:
Here it is in PHP:
```php
$cmds = array('dummy','run','steady');

$encoded = json_encode($cmds);

// do something that sends a GET request to http://<server>/gobra?data= $encoded
```

Basically, Gobra is an API that has 1 endpoint at /gobra
that takes 1 argument `data`.
This argument `data` should be a JSON object that contains `cmds` and `flags`. Both of which are arrays of strings.

Here's an example in a pseudolanguage:
```
data = {
   "cmds": ["dummy", "version"],
   "flags": ["--ok=1"]
}

requestToServer("http://127.0.0.1/gobra?data=" + data.toString())
```


in CURL:
```
curl -X GET 'http://localhost:5000/gobra?data={cmds:["dummy"],flags:[]}'
```

ctessum commented:
Great. Can you add this documentation to the README or somewhere else as appropriate?

> Gobra is an API that has 1 endpoint at /gobra that takes 1 argument data. This argument data should be a JSON object that contains cmds and flags. Both of which are arrays of strings.

ctessum commented:
Also, is it possible to make the endpoint be the name of the command, or configurable by the user, instead of "/gobra"?

khoin commented:
It is possible, and that will be the next commit(s). I'll make changes for this PR and update the documentation. After that, I'll work on the things mentioned above.

khoin commented:
Changes applied

khoin commented:
Question:
> Also, is it possible to make the endpoint be the name of the command, or configurable by the user, instead of "/gobra"?

Do you mean it would end up something like this?
> //some.website/dummy/version
> //127.0.0.1/dummy/run/steady
> //localhost/dummy/addition/?num1=1&num2=30

khoin commented:
I updated the documentation in the last commit to this branch (and squashed it). I wouldn't, however, include the documentation about `?data` and that it has to be a JSON, etc. Since you suggested that it should look like `/command/subcommand/sub2/?flag=value`, I thought that would look much nicer and will later on not reflect with the documentation. 

I would document it when it's at a reasonably stable state.

ctessum commented:
I was mainly just thinking to have it be something other than "gobra", but your proposal sounds like a good idea if it is feasible.
